### ISSUE TITLE
u64 to f32 conversion is inaccurate in Windows 32bit

### ISSUE THREAD

**Issue by icedrocket** [llvm/llvm-project#59787]
On Windows, the x87 control word has a different default value than on other platforms. This not only introduces floating point precision issues, but also results in inconsistent behavior with other platforms. I tried to avoid using x87 by adding the `-msse2` option when compiling for Windows 32bit, but the compiler still generates binaries using only x87.

The code below outputs differently on Windows 32bit than on other platforms:
```c
#include <stdio.h>
#include <stdint.h>

int main() {
    int64_t n = 0b0000000000111111111111111111111111011111111111111111111111111111;
    printf("%lld, %.0f, %.0f", n, (float)n, (float)(uint64_t)n);

    return 0;
}
```

rust-lang/rust#105626

**Comment by topperc** [comment]
Here's a patch to use library call for Windows instead of the broken inline expansion.

```
diff --git a/llvm/lib/Target/X86/X86ISelLowering.cpp b/llvm/lib/Target/X86/X86ISelLowering.cpp
index c0aeed8e45fb..adf236d5d7d0 100644
--- a/llvm/lib/Target/X86/X86ISelLowering.cpp
+++ b/llvm/lib/Target/X86/X86ISelLowering.cpp
@@ -21880,6 +21880,11 @@ SDValue X86TargetLowering::LowerUINT_TO_FP(SDValue Op,
   }
 
   assert(SrcVT == MVT::i64 && "Unexpected type in UINT_TO_FP");
+
+  // Windows default precision control does not support 80-bit FP.
+  if (Subtarget.isOSWindows())
+    return SDValue();
+
   SDValue ValueToStore = Src;
   if (isScalarFPTypeInSSEReg(Op.getValueType()) && !Subtarget.is64Bit()) {
     // Bitcasting to f64 here allows us to do a single 64-bit store from
```

**Comment by icedrocket** [comment]
On Windows 32bit, I also tried `clang.exe test.c -mno-x87 -msse -msse2` but it still only uses x87.

**Comment by icedrocket** [comment]
I also tried `clang.exe test.c -mno-x87 -msse -msse2` but it still only uses x87. I tested it with LLVM for Windows 32bit.

**Comment by Alcaro** [comment]
> I also tried clang.exe test.c -mno-x87 -msse -msse2 but it still only uses x87. I tested it with LLVM for Windows 32bit.

-mfpmath=sse

No clue why -msse/-msse2 doesn't imply that. It doesn't affect ABI, it still passes arguments on stack and return in x87.

**Comment by topperc** [comment]
> On Windows 32bit, I also tried `clang.exe test.c -mno-x87 -msse -msse2` but it still only uses x87.

-mno-x87 is not correctly implemented and is poorly tested. It was added after a lot of code was written and a lot of that hasn’t been updated.

There’s no sse2 instruction that can do unsigned 64-bit integer to float on a 32 bit target. So we use x87. But the algorithm only works if x87 is configured for full precision. If not it rounds incorrectly.

**Comment by phoebewang** [comment]
MSVC has the same problem: https://godbolt.org/z/4sbn8dq1T

**Comment by icedrocket** [comment]
I made the patch below and tested it to see if it works.
```diff
diff --git a/llvm/lib/Target/X86/X86ISelLowering.cpp b/llvm/lib/Target/X86/X86ISelLowering.cpp
index c0aeed8e45fb..d29cf8d48cbc 100644
--- a/llvm/lib/Target/X86/X86ISelLowering.cpp
+++ b/llvm/lib/Target/X86/X86ISelLowering.cpp
@@ -21368,6 +21368,10 @@ SDValue X86TargetLowering::LowerSINT_TO_FP(SDValue Op,
   if (VT == MVT::f128 || !Subtarget.hasX87())
     return SDValue();
 
+  // Windows default precision control does not support 80-bit FP.
+  if (SrcVT == MVT::i64 && Subtarget.isOSWindows())
+    return SDValue();
+
   SDValue ValueToStore = Src;
   if (SrcVT == MVT::i64 && Subtarget.hasSSE2() && !Subtarget.is64Bit())
     // Bitcasting to f64 here allows us to do a single 64-bit store from
@@ -21859,6 +21863,9 @@ SDValue X86TargetLowering::LowerUINT_TO_FP(SDValue Op,
       (DstVT == MVT::f32 || DstVT == MVT::f64))
     return SDValue();
 
+  if (!Subtarget.hasX87())
+    return SDValue();
+
   // Make a 64-bit buffer, and use it to build an FILD.
   SDValue StackSlot = DAG.CreateStackTemporary(MVT::i64, 8);
   int SSFI = cast<FrameIndexSDNode>(StackSlot)->getIndex();
@@ -21880,6 +21887,11 @@ SDValue X86TargetLowering::LowerUINT_TO_FP(SDValue Op,
   }
 
   assert(SrcVT == MVT::i64 && "Unexpected type in UINT_TO_FP");
+
+  // Windows default precision control does not support 80-bit FP.
+  if (Subtarget.isOSWindows())
+    return SDValue();
+
   SDValue ValueToStore = Src;
   if (isScalarFPTypeInSSEReg(Op.getValueType()) && !Subtarget.is64Bit()) {
     // Bitcasting to f64 here allows us to do a single 64-bit store from
```

Unfortunately, it failed to create the executable due to the error below, but it seems to have been replaced with a library call. I'm not sure if implementations of these symbols are provided by LLVM or if I'm missing something in the compile settings.
```
test-b2771c.o : error LNK2019: unresolved external symbol ___floatundisf referenced in function _main
test-b2771c.o : error LNK2019: unresolved external symbol ___floatdisf referenced in function _main
a.exe : fatal error LNK1120: 2 unresolved externals
clang: error: linker command failed with exit code 1120 (use -v to see invocation)
```

**Comment by icedrocket** [comment]
I made the patch below and tested it to see if it works.
```diff
diff --git a/llvm/lib/Target/X86/X86ISelLowering.cpp b/llvm/lib/Target/X86/X86ISelLowering.cpp
index c0aeed8e45fb..d29cf8d48cbc 100644
--- a/llvm/lib/Target/X86/X86ISelLowering.cpp
+++ b/llvm/lib/Target/X86/X86ISelLowering.cpp
@@ -21368,6 +21368,10 @@ SDValue X86TargetLowering::LowerSINT_TO_FP(SDValue Op,
   if (VT == MVT::f128 || !Subtarget.hasX87())
     return SDValue();
 
+  // Windows default precision control does not support 80-bit FP.
+  if (SrcVT == MVT::i64 && Subtarget.isOSWindows())
+    return SDValue();
+
   SDValue ValueToStore = Src;
   if (SrcVT == MVT::i64 && Subtarget.hasSSE2() && !Subtarget.is64Bit())
     // Bitcasting to f64 here allows us to do a single 64-bit store from
@@ -21859,6 +21863,9 @@ SDValue X86TargetLowering::LowerUINT_TO_FP(SDValue Op,
       (DstVT == MVT::f32 || DstVT == MVT::f64))
     return SDValue();
 
+  if (!Subtarget.hasX87())
+    return SDValue();
+
   // Make a 64-bit buffer, and use it to build an FILD.
   SDValue StackSlot = DAG.CreateStackTemporary(MVT::i64, 8);
   int SSFI = cast<FrameIndexSDNode>(StackSlot)->getIndex();
@@ -21880,6 +21887,11 @@ SDValue X86TargetLowering::LowerUINT_TO_FP(SDValue Op,
   }
 
   assert(SrcVT == MVT::i64 && "Unexpected type in UINT_TO_FP");
+
+  // Windows default precision control does not support 80-bit FP.
+  if (Subtarget.isOSWindows())
+    return SDValue();
+
   SDValue ValueToStore = Src;
   if (isScalarFPTypeInSSEReg(Op.getValueType()) && !Subtarget.is64Bit()) {
     // Bitcasting to f64 here allows us to do a single 64-bit store from
```

Unfortunately, the compiler failed to generate the executable due to the errors below, but it seems to be generating code that makes library calls instead of using x87. I'm not sure if implementations of these symbols are provided by LLVM or if I'm missing something in the cmake configuration.
```
test-b2771c.o : error LNK2019: unresolved external symbol ___floatundisf referenced in function _main
test-b2771c.o : error LNK2019: unresolved external symbol ___floatdisf referenced in function _main
a.exe : fatal error LNK1120: 2 unresolved externals
clang: error: linker command failed with exit code 1120 (use -v to see invocation)
```

**Comment by icedrocket** [comment]
I added the compiler-rt option and rebuilt the compiler and the test code compiled successfully with the `--rtlib=compiler-rt` option. When compiled using only x87, it prints the correct value, but when SSE is added, it prints 0.

**Comment by icedrocket** [comment]
I added the compiler-rt option and rebuilt the compiler and the test code compiled successfully with the `--rtlib=compiler-rt` option. The executable outputs the correct value, but outputs 0 if the `-mno-x87` option is added.

**Comment by icedrocket** [comment]
> MSVC has the same problem: https://godbolt.org/z/4sbn8dq1T

MSVC does not seem to have different results depending on whether or not it is signed. It might be better to match MSVC's behavior than to use library calls, even though we lose very little precision in converting between 64-bit integers and floating point.

**Comment by icedrocket** [comment]
> MSVC has the same problem: https://godbolt.org/z/4sbn8dq1T

MSVC x86 does not seem to produce different results depending on whether or not it is signed, so it is less likely to be an issue. It might be better to just match MSVC's behavior than to use library calls, even though we lose a tiny bit of precision.

**Comment by icedrocket** [comment]
Fixed in a6e3027db7ebe6863e44bafcfeaacc16bdc88a3f

**Comment by topperc** [comment]
I'm going to revert the patch. It broke Chrome and Halide builds. I'm going to look at modifying PC in FPCW on the fly.

**Comment by icedrocket** [comment]
Now fixed again in 11fb09ec0afa8620bba44f374763e2ad64332e2e
